File: development/frappe-bench/apps/university_erp/university_erp/university_erp/page/university_dashboard/university_dashboard.js
frappe.pages['university-dashboard'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __('University Dashboard'),
        single_column: true
    });

    new UniversityDashboard(page);
};

class UniversityDashboard {
    constructor(page) {
        this.page = page;
        this.make();
    }

    make() {
        this.$container = $('<div class="university-dashboard p-3"></div>').appendTo(this.page.body);

        // Add refresh button
        this.page.set_primary_action(__('Refresh'), () => this.load_data(), 'refresh');

        this.load_data();
    }

    load_data() {
        this.$container.html(`
            <div class="text-center py-5">
                <div class="spinner-border text-primary" role="status">
                    <span class="visually-hidden">${__('Loading...')}</span>
                </div>
                <p class="mt-2 text-muted">${__('Loading dashboard data...')}</p>
            </div>
        `);

        frappe.call({
            method: 'university_erp.university_erp.page.university_dashboard.university_dashboard.get_dashboard_data',
            callback: (r) => {
                if (r.message) {
                    this.render(r.message);
                }
            },
            error: () => {
                this.$container.html(`
                    <div class="alert alert-danger">
                        ${__('Error loading dashboard data. Please try again.')}
                    </div>
                `);
            }
        });
    }

    render(data) {
        this.$container.html(`
            <!-- Summary Cards Row 1 -->
            <div class="row mb-4">
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="dashboard-stat-card bg-primary text-white">
                        <div class="stat-icon">
                            <i class="fa fa-users"></i>
                        </div>
                        <div class="stat-content">
                            <h3 class="stat-number">${data.students.total || 0}</h3>
                            <p class="stat-label">${__('Total Students')}</p>
                            <small>${__('New this month')}: ${data.students.new_this_month || 0}</small>
                        </div>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="dashboard-stat-card bg-success text-white">
                        <div class="stat-icon">
                            <i class="fa fa-chalkboard-teacher"></i>
                        </div>
                        <div class="stat-content">
                            <h3 class="stat-number">${data.hr.faculty_count || 0}</h3>
                            <p class="stat-label">${__('Faculty Members')}</p>
                            <small>${__('Staff')}: ${data.hr.staff_count || 0}</small>
                        </div>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="dashboard-stat-card bg-info text-white">
                        <div class="stat-icon">
                            <i class="fa fa-rupee-sign"></i>
                        </div>
                        <div class="stat-content">
                            <h3 class="stat-number">${data.finance.collection_rate}%</h3>
                            <p class="stat-label">${__('Fee Collection Rate')}</p>
                            <small>${__('Outstanding')}: ₹${this.formatCurrency(data.finance.total_outstanding)}</small>
                        </div>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="dashboard-stat-card bg-warning text-white">
                        <div class="stat-icon">
                            <i class="fa fa-briefcase"></i>
                        </div>
                        <div class="stat-content">
                            <h3 class="stat-number">${data.placements.students_placed || 0}</h3>
                            <p class="stat-label">${__('Students Placed')}</p>
                            <small>${__('Avg Package')}: ₹${data.placements.avg_package || 0} LPA</small>
                        </div>
                    </div>
                </div>
            </div>

            <!-- Summary Cards Row 2 -->
            <div class="row mb-4">
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="dashboard-stat-card bg-secondary text-white">
                        <div class="stat-icon">
                            <i class="fa fa-book"></i>
                        </div>
                        <div class="stat-content">
                            <h3 class="stat-number">${data.academics.active_courses || 0}</h3>
                            <p class="stat-label">${__('Active Courses')}</p>
                            <small>${__('Attendance')}: ${data.academics.avg_attendance}%</small>
                        </div>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="dashboard-stat-card bg-dark text-white">
                        <div class="stat-icon">
                            <i class="fa fa-building"></i>
                        </div>
                        <div class="stat-content">
                            <h3 class="stat-number">${data.hostel.occupancy_rate}%</h3>
                            <p class="stat-label">${__('Hostel Occupancy')}</p>
                            <small>${__('Available')}: ${data.hostel.available || 0} beds</small>
                        </div>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="dashboard-stat-card" style="background: #6f42c1; color: white;">
                        <div class="stat-icon">
                            <i class="fa fa-book-reader"></i>
                        </div>
                        <div class="stat-content">
                            <h3 class="stat-number">${data.library.books_issued || 0}</h3>
                            <p class="stat-label">${__('Books Issued')}</p>
                            <small>${__('Overdue')}: ${data.library.overdue_books || 0}</small>
                        </div>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="dashboard-stat-card" style="background: #fd7e14; color: white;">
                        <div class="stat-icon">
                            <i class="fa fa-user-graduate"></i>
                        </div>
                        <div class="stat-content">
                            <h3 class="stat-number">${data.admissions.total_applications || 0}</h3>
                            <p class="stat-label">${__('Applications')}</p>
                            <small>${__('Pending')}: ${data.admissions.pending || 0}</small>
                        </div>
                    </div>
                </div>
            </div>

            <!-- Charts Row -->
            <div class="row mb-4">
                <div class="col-md-6 mb-3">
                    <div class="chart-container card">
                        <div class="card-header">
                            <h6 class="mb-0">${__('Fee Collection Trend')}</h6>
                        </div>
                        <div class="card-body">
                            <div id="fee-chart" style="height: 250px;"></div>
                        </div>
                    </div>
                </div>
                <div class="col-md-6 mb-3">
                    <div class="chart-container card">
                        <div class="card-header">
                            <h6 class="mb-0">${__('Students by Program')}</h6>
                        </div>
                        <div class="card-body">
                            <div id="program-chart" style="height: 250px;"></div>
                        </div>
                    </div>
                </div>
            </div>

            <div class="row mb-4">
                <div class="col-md-6 mb-3">
                    <div class="chart-container card">
                        <div class="card-header">
                            <h6 class="mb-0">${__('Faculty by Department')}</h6>
                        </div>
                        <div class="card-body">
                            <div id="dept-chart" style="height: 250px;"></div>
                        </div>
                    </div>
                </div>
                <div class="col-md-6 mb-3">
                    <div class="chart-container card">
                        <div class="card-header">
                            <h6 class="mb-0">${__('Admission Status')}</h6>
                        </div>
                        <div class="card-body">
                            <div id="admission-chart" style="height: 250px;"></div>
                        </div>
                    </div>
                </div>
            </div>

            <!-- Quick Actions -->
            <div class="row">
                <div class="col-12">
                    <div class="card">
                        <div class="card-header">
                            <h6 class="mb-0">${__('Quick Actions')}</h6>
                        </div>
                        <div class="card-body">
                            <div class="row">
                                <div class="col-md-2 col-sm-4 col-6 mb-2 text-center">
                                    <a href="/app/student" class="btn btn-outline-primary btn-block">
                                        <i class="fa fa-users d-block mb-1"></i>
                                        ${__('Students')}
                                    </a>
                                </div>
                                <div class="col-md-2 col-sm-4 col-6 mb-2 text-center">
                                    <a href="/app/fees" class="btn btn-outline-success btn-block">
                                        <i class="fa fa-rupee-sign d-block mb-1"></i>
                                        ${__('Fees')}
                                    </a>
                                </div>
                                <div class="col-md-2 col-sm-4 col-6 mb-2 text-center">
                                    <a href="/app/teaching-assignment" class="btn btn-outline-info btn-block">
                                        <i class="fa fa-chalkboard d-block mb-1"></i>
                                        ${__('Assignments')}
                                    </a>
                                </div>
                                <div class="col-md-2 col-sm-4 col-6 mb-2 text-center">
                                    <a href="/app/placement-drive" class="btn btn-outline-warning btn-block">
                                        <i class="fa fa-briefcase d-block mb-1"></i>
                                        ${__('Placements')}
                                    </a>
                                </div>
                                <div class="col-md-2 col-sm-4 col-6 mb-2 text-center">
                                    <a href="/app/hostel-allocation" class="btn btn-outline-secondary btn-block">
                                        <i class="fa fa-building d-block mb-1"></i>
                                        ${__('Hostel')}
                                    </a>
                                </div>
                                <div class="col-md-2 col-sm-4 col-6 mb-2 text-center">
                                    <a href="/app/library-transaction" class="btn btn-outline-dark btn-block">
                                        <i class="fa fa-book d-block mb-1"></i>
                                        ${__('Library')}
                                    </a>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        `);

        this.render_charts(data);
    }

    render_charts(data) {
        // Fee Collection Trend
        if (data.finance.monthly_trend && data.finance.monthly_trend.length) {
            new frappe.Chart("#fee-chart", {
                data: {
                    labels: data.finance.monthly_trend.map(d => d.month),
                    datasets: [{
                        name: __("Collected"),
                        values: data.finance.monthly_trend.map(d => d.collected || 0)
                    }]
                },
                type: 'line',
                colors: ['#5e64ff'],
                lineOptions: {
                    regionFill: 1
                },
                axisOptions: {
                    xIsSeries: true
                }
            });
        } else {
            $('#fee-chart').html(`<p class="text-muted text-center py-5">${__('No fee data available')}</p>`);
        }

        // Students by Program
        if (data.students.by_program && data.students.by_program.length) {
            new frappe.Chart("#program-chart", {
                data: {
                    labels: data.students.by_program.map(d => (d.program || 'Unknown').substring(0, 20)),
                    datasets: [{
                        name: __("Students"),
                        values: data.students.by_program.map(d => d.count)
                    }]
                },
                type: 'bar',
                colors: ['#28a745'],
                barOptions: {
                    spaceRatio: 0.3
                }
            });
        } else {
            $('#program-chart').html(`<p class="text-muted text-center py-5">${__('No program data available')}</p>`);
        }

        // Faculty by Department
        if (data.hr.by_department && data.hr.by_department.length) {
            new frappe.Chart("#dept-chart", {
                data: {
                    labels: data.hr.by_department.map(d => (d.department || 'N/A').substring(0, 15)),
                    datasets: [{
                        name: __("Faculty"),
                        values: data.hr.by_department.map(d => d.count)
                    }]
                },
                type: 'pie',
                colors: ['#5e64ff', '#28a745', '#ffc107', '#dc3545', '#17a2b8', '#6c757d', '#fd7e14', '#6f42c1']
            });
        } else {
            $('#dept-chart').html(`<p class="text-muted text-center py-5">${__('No department data available')}</p>`);
        }

        // Admission Status
        if (data.admissions.total_applications > 0) {
            new frappe.Chart("#admission-chart", {
                data: {
                    labels: [__('Pending'), __('Approved'), __('Admitted'), __('Rejected')],
                    datasets: [{
                        name: __("Applications"),
                        values: [
                            data.admissions.pending || 0,
                            data.admissions.approved || 0,
                            data.admissions.admitted || 0,
                            data.admissions.rejected || 0
                        ]
                    }]
                },
                type: 'donut',
                colors: ['#ffc107', '#28a745', '#5e64ff', '#dc3545']
            });
        } else {
            $('#admission-chart').html(`<p class="text-muted text-center py-5">${__('No admission data available')}</p>`);
        }
    }

    formatCurrency(value) {
        if (!value) return '0';
        if (value >= 10000000) {
            return (value / 10000000).toFixed(2) + ' Cr';
        } else if (value >= 100000) {
            return (value / 100000).toFixed(2) + ' L';
        } else if (value >= 1000) {
            return (value / 1000).toFixed(2) + ' K';
        }
        return value.toFixed(2);
    }
}

File: development/frappe-bench/apps/university_erp/university_erp/university_erp/page/university_dashboard/university_dashboard.py
# Copyright (c) 2026, University ERP and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import nowdate, add_months, getdate, flt


@frappe.whitelist()
def get_dashboard_data():
    """Get comprehensive university dashboard data"""
    data = {
        "students": get_student_stats(),
        "academics": get_academic_stats(),
        "finance": get_finance_stats(),
        "hr": get_hr_stats(),
        "admissions": get_admission_stats(),
        "placements": get_placement_stats(),
        "hostel": get_hostel_stats(),
        "library": get_library_stats()
    }

    return data


def get_student_stats():
    """Student-related statistics"""
    total = frappe.db.count("Student", {"enabled": 1})

    by_program = frappe.db.sql("""
        SELECT program, COUNT(*) as count
        FROM `tabStudent`
        WHERE enabled = 1 AND program IS NOT NULL
        GROUP BY program
        ORDER BY count DESC
        LIMIT 10
    """, as_dict=True)

    by_gender = frappe.db.sql("""
        SELECT gender, COUNT(*) as count
        FROM `tabStudent`
        WHERE enabled = 1 AND gender IS NOT NULL
        GROUP BY gender
    """, as_dict=True)

    # New enrollments this month
    new_this_month = frappe.db.count("Student", {
        "enabled": 1,
        "creation": [">=", add_months(nowdate(), -1)]
    })

    return {
        "total": total,
        "by_program": by_program,
        "by_gender": by_gender,
        "new_this_month": new_this_month
    }


def get_academic_stats():
    """Academic statistics"""
    # Get current academic term
    current_term = frappe.db.get_value(
        "Academic Term",
        {"term_start_date": ["<=", nowdate()], "term_end_date": [">=", nowdate()]},
        "name"
    )

    # Count active courses
    active_courses = frappe.db.count("Course", {"is_published": 1}) if frappe.db.has_column("Course", "is_published") else frappe.db.count("Course")

    # Count programs
    total_programs = frappe.db.count("Program")

    # Average attendance (last 30 days)
    avg_attendance = frappe.db.sql("""
        SELECT AVG(
            CASE WHEN status = 'Present' THEN 100 ELSE 0 END
        ) as avg
        FROM `tabStudent Attendance`
        WHERE date >= %s
    """, (add_months(nowdate(), -1),))[0][0] or 0

    # Teaching assignments count
    teaching_assignments = frappe.db.count("Teaching Assignment", {"docstatus": 1})

    return {
        "current_term": current_term,
        "active_courses": active_courses,
        "total_programs": total_programs,
        "avg_attendance": round(avg_attendance, 1),
        "teaching_assignments": teaching_assignments
    }


def get_finance_stats():
    """Financial statistics"""
    current_fy = get_current_fiscal_year()
    start_date = current_fy.get("start_date") if current_fy else add_months(nowdate(), -12)

    fee_stats = frappe.db.sql("""
        SELECT
            COALESCE(SUM(grand_total), 0) as total_generated,
            COALESCE(SUM(grand_total - outstanding_amount), 0) as total_collected,
            COALESCE(SUM(outstanding_amount), 0) as total_outstanding
        FROM `tabFees`
        WHERE docstatus = 1
        AND posting_date >= %s
    """, (start_date,), as_dict=True)[0]

    # Monthly collection trend
    monthly_collection = frappe.db.sql("""
        SELECT
            DATE_FORMAT(posting_date, '%%Y-%%m') as month,
            SUM(grand_total - outstanding_amount) as collected
        FROM `tabFees`
        WHERE docstatus = 1
        AND posting_date >= %s
        GROUP BY DATE_FORMAT(posting_date, '%%Y-%%m')
        ORDER BY month
        LIMIT 12
    """, (add_months(nowdate(), -12),), as_dict=True)

    collection_rate = 0
    if fee_stats.get("total_generated"):
        collection_rate = round(
            (fee_stats.get("total_collected") or 0) /
            fee_stats.get("total_generated") * 100, 1
        )

    return {
        "total_generated": flt(fee_stats.get("total_generated")),
        "total_collected": flt(fee_stats.get("total_collected")),
        "total_outstanding": flt(fee_stats.get("total_outstanding")),
        "collection_rate": collection_rate,
        "monthly_trend": monthly_collection
    }


def get_hr_stats():
    """HR statistics"""
    # Total employees
    total_employees = frappe.db.count("Employee", {"status": "Active"})

    # Faculty count
    faculty_count = frappe.db.count(
        "Employee",
        {"custom_employee_category": "Teaching", "status": "Active"}
    ) if frappe.db.has_column("Employee", "custom_employee_category") else 0

    # Staff count
    staff_count = total_employees - faculty_count

    # Faculty by department
    by_department = frappe.db.sql("""
        SELECT
            COALESCE(custom_academic_department, 'Unassigned') as department,
            COUNT(*) as count
        FROM `tabEmployee`
        WHERE status = 'Active'
        AND custom_employee_category = 'Teaching'
        GROUP BY custom_academic_department
        ORDER BY count DESC
        LIMIT 10
    """, as_dict=True) if frappe.db.has_column("Employee", "custom_academic_department") else []

    # Pending leave applications
    pending_leaves = frappe.db.count("Leave Application", {"status": "Open"})

    return {
        "total_employees": total_employees,
        "faculty_count": faculty_count,
        "staff_count": staff_count,
        "by_department": by_department,
        "pending_leaves": pending_leaves
    }


def get_admission_stats():
    """Admission statistics"""
    # Get current academic year
    current_ay = frappe.db.get_value(
        "Academic Year",
        {"year_start_date": ["<=", nowdate()], "year_end_date": [">=", nowdate()]},
        "name"
    )

    if not current_ay:
        return {
            "academic_year": None,
            "total_applications": 0,
            "approved": 0,
            "admitted": 0,
            "pending": 0,
            "rejected": 0
        }

    applications = frappe.db.sql("""
        SELECT
            COUNT(*) as total,
            SUM(CASE WHEN application_status = 'Approved' THEN 1 ELSE 0 END) as approved,
            SUM(CASE WHEN application_status = 'Admitted' THEN 1 ELSE 0 END) as admitted,
            SUM(CASE WHEN application_status IN ('Applied', 'Pending') THEN 1 ELSE 0 END) as pending,
            SUM(CASE WHEN application_status = 'Rejected' THEN 1 ELSE 0 END) as rejected
        FROM `tabStudent Applicant`
        WHERE academic_year = %s
    """, (current_ay,), as_dict=True)[0]

    return {
        "academic_year": current_ay,
        "total_applications": applications.get("total") or 0,
        "approved": applications.get("approved") or 0,
        "admitted": applications.get("admitted") or 0,
        "pending": applications.get("pending") or 0,
        "rejected": applications.get("rejected") or 0
    }


def get_placement_stats():
    """Placement statistics"""
    # Students placed
    placed = frappe.db.count(
        "Placement Application",
        {"status": "Offer Accepted", "docstatus": ["!=", 2]}
    )

    # Active companies
    companies = frappe.db.count("Placement Company")

    # Average package
    avg_package = frappe.db.sql("""
        SELECT AVG(pjo.package_offered) as avg
        FROM `tabPlacement Application` pa
        JOIN `tabPlacement Job Opening` pjo ON pa.job_opening = pjo.name
        WHERE pa.status = 'Offer Accepted'
    """)[0][0] or 0

    # Highest package
    highest_package = frappe.db.sql("""
        SELECT MAX(pjo.package_offered) as max
        FROM `tabPlacement Application` pa
        JOIN `tabPlacement Job Opening` pjo ON pa.job_opening = pjo.name
        WHERE pa.status = 'Offer Accepted'
    """)[0][0] or 0

    # Active job openings
    active_openings = frappe.db.count("Placement Job Opening", {"status": "Open"})

    return {
        "students_placed": placed,
        "active_companies": companies,
        "avg_package": round(avg_package, 2),
        "highest_package": round(highest_package, 2),
        "active_openings": active_openings
    }


def get_hostel_stats():
    """Hostel statistics"""
    # Total capacity
    total_capacity = frappe.db.sql("""
        SELECT COALESCE(SUM(capacity), 0) as total
        FROM `tabHostel Room`
    """)[0][0] or 0

    # Occupied beds
    occupied = frappe.db.sql("""
        SELECT COALESCE(SUM(occupied_beds), 0) as total
        FROM `tabHostel Room`
    """)[0][0] or 0

    # Occupancy rate
    occupancy_rate = round(occupied / total_capacity * 100, 1) if total_capacity else 0

    # Active allocations
    active_allocations = frappe.db.count("Hostel Allocation", {"status": "Active"})

    return {
        "total_capacity": total_capacity,
        "occupied": occupied,
        "available": total_capacity - occupied,
        "occupancy_rate": occupancy_rate,
        "active_allocations": active_allocations
    }


def get_library_stats():
    """Library statistics"""
    # Total articles
    total_articles = frappe.db.count("Library Article")

    # Available articles
    available = frappe.db.sql("""
        SELECT COALESCE(SUM(available_count), 0) as total
        FROM `tabLibrary Article`
    """)[0][0] or 0

    # Active members
    active_members = frappe.db.count("Library Member", {"status": "Active"})

    # Books issued (active transactions)
    books_issued = frappe.db.sql("""
        SELECT COUNT(*) FROM `tabLibrary Transaction` lt
        WHERE lt.type = 'Issue'
        AND lt.docstatus = 1
        AND lt.name NOT IN (
            SELECT issue_reference FROM `tabLibrary Transaction`
            WHERE type = 'Return' AND docstatus = 1 AND issue_reference IS NOT NULL
        )
    """)[0][0] or 0

    # Overdue books
    overdue = frappe.db.sql("""
        SELECT COUNT(*) FROM `tabLibrary Transaction` lt
        WHERE lt.type = 'Issue'
        AND lt.docstatus = 1
        AND lt.to_date < %s
        AND lt.name NOT IN (
            SELECT issue_reference FROM `tabLibrary Transaction`
            WHERE type = 'Return' AND docstatus = 1 AND issue_reference IS NOT NULL
        )
    """, (nowdate(),))[0][0] or 0

    return {
        "total_articles": total_articles,
        "available": available,
        "active_members": active_members,
        "books_issued": books_issued,
        "overdue_books": overdue
    }


def get_current_fiscal_year():
    """Get current fiscal year"""
    fy = frappe.db.get_value(
        "Fiscal Year",
        {"year_start_date": ["<=", nowdate()], "year_end_date": [">=", nowdate()]},
        ["year_start_date", "year_end_date"],
        as_dict=True
    )
    return fy or {"start_date": add_months(nowdate(), -12), "end_date": nowdate()}


@frappe.whitelist()
def get_quick_stats():
    """Get quick stats for dashboard header"""
    return {
        "students": frappe.db.count("Student", {"enabled": 1}),
        "faculty": frappe.db.count("Employee", {"status": "Active", "custom_employee_category": "Teaching"}) if frappe.db.has_column("Employee", "custom_employee_category") else 0,
        "courses": frappe.db.count("Course"),
        "programs": frappe.db.count("Program")
    }
